pkg/istio-agent/xds_proxy.go

```diff
@@ -664,84 +660,37 @@ func (p *XdsProxy) initIstiodDialOptions(agent *Agent) error {
-	initialConnWindowSizeOption := grpc.WithInitialConnWindowSize(int32(defaultInitialConnWindowSize))
-	msgSizeOption := grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(defaultClientMaxReceiveMessageSize))
-	dialOptions := []grpc.DialOption{
-		tlsOpts,
-		keepaliveOption, initialWindowSizeOption, initialConnWindowSizeOption, msgSizeOption,
+	options, err := istiogrpc.ClientOptions(nil, tlsOpts)
+	if err != nil {
+		return nil, err
 	}
-
 	if sa.secOpts.CredFetcher != nil {
-		dialOptions = append(dialOptions, grpc.WithPerRPCCredentials(caclient.NewXDSTokenProvider(sa.secOpts)))
+		options = append(options, grpc.WithPerRPCCredentials(caclient.NewXDSTokenProvider(sa.secOpts)))
 	}
-	return dialOptions, nil
+	return options, nil
 }
 
 // Returns the TLS option to use when talking to Istiod
-// If provisioned cert is set, it will return a mTLS related config
-// Else it will return a one-way TLS related config with the assumption
-// that the consumer code will use tokens to authenticate the upstream.
-func (p *XdsProxy) getTLSDialOption(agent *Agent) (grpc.DialOption, error) {
+func (p *XdsProxy) getTLSOptions(agent *Agent) (*istiogrpc.TLSOptions, error) {
 	if agent.proxyConfig.ControlPlaneAuthPolicy == meshconfig.AuthenticationPolicy_NONE {
-		return grpc.WithTransportCredentials(insecure.NewCredentials()), nil
+		return nil, nil
 	}
-	rootCert, err := p.getRootCertificate(agent)
+	xdsCACertPath, err := agent.FindRootCAForXDS()
 	if err != nil {
-		return nil, err
-	}
-
-	config := tls.Config{
-		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
-			var certificate tls.Certificate
-			key, cert := agent.GetKeyCertsForXDS()
-			if key != "" && cert != "" {
-				var isExpired bool
-				isExpired, err = util.IsCertExpired(cert)
-				if err != nil {
-					log.Warnf("cannot parse the cert chain, using token instead: %v", err)
-					return &certificate, nil
-				}
-				if isExpired {
-					log.Warnf("cert expired, using token instead")
-					return &certificate, nil
-				}
-				// Load the certificate from disk
-				certificate, err = tls.LoadX509KeyPair(cert, key)
-				if err != nil {
-					return nil, err
-				}
-			}
-			return &certificate, nil
-		},
-		RootCAs:    rootCert,
-		MinVersion: tls.VersionTLS12,
-	}
-
-	if host, _, err := net.SplitHostPort(agent.proxyConfig.DiscoveryAddress); err == nil {
-		config.ServerName = host
-	}
-	// For debugging on localhost (with port forward)
-	// This matches the logic for the CA; this code should eventually be shared
-	if strings.Contains(config.ServerName, "localhost") {
-		config.ServerName = "istiod.istio-system.svc"
-	}
-
-	if p.istiodSAN != "" {
-		config.ServerName = p.istiodSAN
+		return nil, fmt.Errorf("failed to find root CA cert for XDS: %v", err)
 	}
-	transportCreds := credentials.NewTLS(&config)
-	return grpc.WithTransportCredentials(transportCreds), nil
+	key, cert := agent.GetKeyCertsForXDS()
```

Comment:
This needs to be dynamic, right now we would not refresh?

Reply:
it will.  the key, cert actually are file names